### PR TITLE
Decode IP field data in Gdn_Controller::renderData

### LINKED ISSUE #4590: Binary IP addresses need to be formatted for API requests
Now that IP addresses are formatted in binary, they need to be properly formatted when returning an API result. Calling the format IP function on resulting fields will be sufficient. I'd do this in `Gdn_Controller->renderData()`.

This is actually a high priority bug as binary IP addresses completely kills the JSON. What I'd like to see with this pull request is:

- [x] Reproduce the error on localhost.
- [x] Try getting JSON_PARTIAL_OUTPUT_ON_ERROR to work with a fallback for PHP 5.4. This will protect us in case we miss something.
- [x] Add the appropriate IP address encoding.
- [x] When JSON encoding a result, if that result is an error (false) then ouput an JSON encoded error about the encoding rather than a white screen of death. Note that there are some good functions for this in php 5.5, but we'll need to conditionally use them because we must support php 5.4.

### PR DESCRIPTION
**Pull request by initvector**
This update addresses issues with binary values causing JSON encoding to fail in `Gdn_Controller::RenderData`.  It does this by recursively iterating through the `$Data` array and running `ipDecode` against fields with recognizable names (e.g. *IPAddress, *IPAddresses).

*Note: Elements in the `$Data` array can be other arrays ___or objects___, making this not an ideal candidate for `array_walk`/`array_walk_recursive`.  These functions skip over iterating through object properties.  Also, with `array_walk_recursive`, any key that holds an array will not be passed to the callback function. This makes it tough to handle values like "AllIPAddresses", which is an array, itself.*

Closes #4590

### CODE REVIEW

File: library/core/class.controller.php [hunk 1]
```diff
@@ -1470,6 +1470,24 @@ public function renderData($Data = null) {
...
+                            $ip = ipDecode($ip);
+                        });
+                    } elseif (is_array($val) || is_object($val)) {
+                        $ipWalk($val);
+                    }
+                }
+            }
+        };
+        $ipWalk($Data);
+
```
Comment (tburry) [inline library/core/class.controller.php]: My preference would be for this to be put into a function/method that could be unit tested.

File: library/core/class.controller.php [hunk 2]
```diff
@@ -1074,6 +1074,28 @@ public function json($Key, $Value = null) {
...
+     */
+    private function jsonEncode($value, $options = null) {
+        if ($options === null) {
+            $options = JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES;
+        }
+
+        $encoded = json_encode($value, $options);
+
+        if (json_last_error() !== JSON_ERROR_NONE) {
+            throw new Gdn_UserException('A JSON encoding error has occurred');
```
Comment (tburry) [inline library/core/class.controller.php]: - This probably isn't a Gdn_UserException(). It's more of a fatal Exception in nature.
- You are missing an HTTP compatible error code (probably a 5xx error).
- We also usually make messages sentences with periods.
- You can conditionally show a more useful message on newer versions of PHP.

File: library/core/class.controller.php [hunk 3]
```diff
@@ -1074,6 +1074,28 @@ public function json($Key, $Value = null) {
...
+     * Encode a value as JSON or throw an exception on error.
+     *
+     * @param mixed $value
+     * @param int|null $options
+     * @return string
+     * @throws Gdn_UserException
+     */
+    private function jsonEncode($value, $options = null) {
+        if ($options === null) {
+            $options = JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES;
```
Comment (tburry) [inline library/core/class.controller.php]: Conditionally add supported options from newer PHP (see original issue).

File: library/core/class.controller.php [hunk 4]
```diff
@@ -1074,6 +1074,28 @@ public function json($Key, $Value = null) {
...
 
     /**
+     * Encode a value as JSON or throw an exception on error.
+     *
+     * @param mixed $value
+     * @param int|null $options
+     * @return string
+     * @throws Gdn_UserException
+     */
+    private function jsonEncode($value, $options = null) {
```
Comment (tburry) [inline library/core/class.controller.php]: Make a global function and unit test.

File: library/core/functions.general.php [hunk 5]
```diff
@@ -3846,3 +3863,22 @@ function urlMatch($pattern, $url) {
...
+
+if (!function_exists('walkAll')) {
+    /**
+     * Recursively walk through all array elements or object properties.
+     *
+     * @param array|object $input
+     * @param callable $callback
+     * @param string $parent Array key or object property to indicate parent node.
+     */
+    function walkAll(&$input, $callback, $parent = null) {
```
Comment (tburry) [inline library/core/functions.general.php]: - The standard PHP practice puts the word "recursive" at the end of these such function names.
- I'm not a huge fan of making the $parent parameter something that the original caller can pass. Perhaps an inner closure that exposes it is better.
- Cycles can create infinite recursion. Sad.

File: library/core/functions.general.php [hunk 6]
```diff
@@ -1767,6 +1767,23 @@ function ipDecode($packedIP) {
     }
 }
 
+if (!function_exists('ipDecodeWalk')) {
+    /**
+     * Recursively walk through the data, decoding any IP address fields.
+     *
+     * @param array|object $data
+     */
+    function ipDecodeWalk(&$data) {
```
Comment (tburry) [inline library/core/functions.general.php]: ipDecodeRecursive. This is probably a method that can be left on the controller.

**Comment by initvector** [comment]
I think everything has been addressed.

I added a test for `jsonEncodeChecked`, which basically makes sure it's throwing exceptions to indicate a failure to encode data, instead of silently returning `false`/`null`.

**Pull request merged**